### ISSUE TITLE
torch.compile produces large absolute errors for model with LSTM, MultiheadAttention, and torch.reciprocal

### ISSUE THREAD

**Issue by Blooming-Tree** [pytorch/pytorch#188430]
### 🐛 Describe the bug

When applying torch.compile to a model that combines LSTM, multiple nn.MultiheadAttention layers, and torch.reciprocal operations, the compiled model produces outputs with absolute errors that significantly exceed acceptable thresholds. The maximum absolute difference reaches 1.92, which is nearly 200 times larger than the 0.01 tolerance.
code:
```
import os
import torch
import copy
import torch.nn as nn
import torch.nn.functional as F

class AttentionModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.lstm = nn.LSTM(input_size=5, hidden_size=20, num_layers=1, batch_first=False)
        self.linear1 = nn.Linear(20, 20)
        self.attention1 = nn.MultiheadAttention(embed_dim=20, num_heads=2, batch_first=torch.tensor(False, device='cpu'))
        self.relu = nn.ReLU()
        self.linear2 = nn.Linear(20, 15)
        self.attention2 = nn.MultiheadAttention(embed_dim=15, num_heads=5, batch_first=False)
        self.linear3 = nn.Linear(15, 10)
        self.attention3 = nn.MultiheadAttention(embed_dim=10, num_heads=2, batch_first=False)

    def forward(self, x):
        lstm_out, _ = self.lstm(x)
        linear1_out = self.relu(self.linear1(lstm_out))
        attn1_out, _ = self.attention1(linear1_out, linear1_out, linear1_out)
        reciprocal1 = torch.reciprocal(attn1_out)
        linear2_out = self.relu(self.linear2(reciprocal1))
        attn2_out, _ = self.attention2(linear2_out, linear2_out, linear2_out)
        reciprocal2 = attn2_out.reciprocal()
        linear3_out = self.relu(self.linear3(reciprocal2))
        attn3_out, _ = self.attention3(linear3_out, linear3_out, linear3_out)
        reciprocal3 = torch.reciprocal(attn3_out)
        return reciprocal3

def get_default_model():
    return AttentionModel()

def get_sample_inputs():
    return (torch.randn(10, 2, 5, dtype=torch.float64),)

def main():
    torch.use_deterministic_algorithms(True)
    torch.set_float32_matmul_precision('highest')
    os.environ['CUBLAS_WORKSPACE_CONFIG'] = ':4096:8'
    torch._inductor.config.fallback_random = True

    torch.manual_seed(42)
    model = get_default_model()
    model = model.double()
    model.eval()

    inputs = get_sample_inputs()
    inputs = tuple(t.double() if isinstance(t, torch.Tensor) and t.is_floating_point() else t for t in inputs)

    with torch.no_grad():
        eager_output = model(*inputs)
    print(f'eager float64 output: {eager_output}')

    torch.manual_seed(42)
    model_fp32 = copy.deepcopy(model)
    model_fp32 = model_fp32.float()
    model_fp32.eval()

    inputs_fp32 = tuple(t.float() if isinstance(t, torch.Tensor) and t.is_floating_point() else t for t in inputs)
    compiled_model = torch.compile(model_fp32, mode="reduce-overhead")

    with torch.no_grad():
        compiled_output = compiled_model(*inputs_fp32)
    print(f'compile float32 output: {compiled_output}')

    abs_diff = (eager_output - compiled_output.double()).abs()
    rel_diff = abs_diff / (eager_output.abs() + 1e-8)
    is_inconsistent = (abs_diff > 0.01) | (rel_diff > 0.001)

    print(f'Abs diff max: {abs_diff.max().item():.6e}, mean: {abs_diff.mean().item():.6e}')
    print(f'Rel diff max: {rel_diff.max().item():.6e}, mean: {rel_diff.mean().item():.6e}')
    print(f'Inconsistent: {is_inconsistent.sum().item()} / {eager_output.numel()}')

if __name__ == '__main__':
    main()
```

### Error logs

output:
```
eager float64 output: tensor([[[   20.5618,  -180.4886,  -372.6351,  1946.2752,   816.3757,
          -1231.0008,    13.7859,   128.2286,   -71.1102,   158.7893],
         [   17.4994,  -569.4491,    54.2834,  -571.2079,   135.8356,
            -93.6676,    11.7999,   887.8607,   -41.1327,   467.4725]],

        [[   20.5618,  -180.4886,  -372.6351,  1946.2752,   816.3757,
          -1231.0008,    13.7859,   128.2286,   -71.1102,   158.7893],
         [   17.4994,  -569.4491,    54.2834,  -571.2079,   135.8356,
            -93.6676,    11.7999,   887.8607,   -41.1327,   467.4725]],

        [[   20.5618,  -180.4886,  -372.6351,  1946.2752,   816.3757,
          -1231.0008,    13.7859,   128.2286,   -71.1102,   158.7893],
         [   17.4994,  -569.4491,    54.2834,  -571.2079,   135.8356,
            -93.6676,    11.7999,   887.8607,   -41.1327,   467.4725]],

        [[   20.5618,  -180.4886,  -372.6351,  1946.2752,   816.3757,
          -1231.0008,    13.7859,   128.2286,   -71.1102,   158.7893],
         [   17.4994,  -569.4491,    54.2834,  -571.2079,   135.8356,
            -93.6676,    11.7999,   887.8607,   -41.1327,   467.4725]],

        [[   20.5618,  -180.4886,  -372.6351,  1946.2752,   816.3757,
          -1231.0008,    13.7859,   128.2286,   -71.1102,   158.7893],
         [   17.4994,  -569.4491,    54.2834,  -571.2079,   135.8356,
            -93.6676,    11.7999,   887.8607,   -41.1327,   467.4725]],

        [[   20.5618,  -180.4886,  -372.6351,  1946.2752,   816.3757,
          -1231.0008,    13.7859,   128.2286,   -71.1102,   158.7893],
         [   17.4994,  -569.4491,    54.2834,  -571.2079,   135.8356,
            -93.6676,    11.7999,   887.8607,   -41.1327,   467.4725]],

        [[   20.5618,  -180.4886,  -372.6351,  1946.2752,   816.3757,
          -1231.0008,    13.7859,   128.2286,   -71.1102,   158.7893],
         [   17.4994,  -569.4491,    54.2834,  -571.2079,   135.8356,
            -93.6676,    11.7999,   887.8607,   -41.1327,   467.4725]],

        [[   20.5618,  -180.4886,  -372.6351,  1946.2752,   816.3757,
          -1231.0008,    13.7859,   128.2286,   -71.1102,   158.7893],
         [   17.4994,  -569.4491,    54.2834,  -571.2079,   135.8356,
            -93.6676,    11.7999,   887.8607,   -41.1327,   467.4725]],

        [[   20.5618,  -180.4886,  -372.6351,  1946.2752,   816.3757,
          -1231.0008,    13.7859,   128.2286,   -71.1102,   158.7893],
         [   17.4994,  -569.4491,    54.2834,  -571.2079,   135.8356,
            -93.6676,    11.7999,   887.8607,   -41.1327,   467.4725]],

        [[   20.5618,  -180.4886,  -372.6351,  1946.2752,   816.3757,
          -1231.0008,    13.7859,   128.2286,   -71.1102,   158.7893],
         [   17.4994,  -569.4491,    54.2834,  -571.2079,   135.8356,
            -93.6676,    11.7999,   887.8607,   -41.1327,   467.4725]]],
       dtype=torch.float64)
compile float32 output: tensor([[[   20.5619,  -180.5011,  -372.7271,  1948.1963,   815.7861,
          -1230.3558,    13.7859,   128.2355,   -71.1078,   158.7931],
         [   17.4993,  -569.6699,    54.2848,  -571.0992,   135.8392,
            -93.6641,    11.7999,   887.9709,   -41.1330,   467.5546]],

        [[   20.5619,  -180.5011,  -372.7271,  1948.1963,   815.7861,
          -1230.3558,    13.7859,   128.2355,   -71.1078,   158.7931],
         [   17.4993,  -569.6699,    54.2848,  -571.0992,   135.8392,
            -93.6641,    11.7999,   887.9709,   -41.1330,   467.5546]],

        [[   20.5619,  -180.5011,  -372.7271,  1948.1963,   815.7861,
          -1230.3558,    13.7859,   128.2355,   -71.1078,   158.7931],
         [   17.4993,  -569.6699,    54.2848,  -571.0992,   135.8392,
            -93.6641,    11.7999,   887.9709,   -41.1330,   467.5546]],

        [[   20.5619,  -180.5011,  -372.7271,  1948.1963,   815.7861,
          -1230.3558,    13.7859,   128.2355,   -71.1078,   158.7931],
         [   17.4993,  -569.6699,    54.2848,  -571.0992,   135.8392,
            -93.6641,    11.7999,   887.9709,   -41.1330,   467.5546]],

        [[   20.5619,  -180.5011,  -372.7271,  1948.1963,   815.7861,
          -1230.3558,    13.7859,   128.2355,   -71.1078,   158.7931],
         [   17.4993,  -569.6699,    54.2848,  -571.0992,   135.8392,
            -93.6641,    11.7999,   887.9709,   -41.1330,   467.5546]],

        [[   20.5619,  -180.5011,  -372.7271,  1948.1963,   815.7861,
          -1230.3558,    13.7859,   128.2355,   -71.1078,   158.7931],
         [   17.4993,  -569.6699,    54.2848,  -571.0992,   135.8392,
            -93.6641,    11.7999,   887.9709,   -41.1330,   467.5546]],

        [[   20.5619,  -180.5011,  -372.7271,  1948.1963,   815.7861,
          -1230.3558,    13.7859,   128.2355,   -71.1078,   158.7931],
         [   17.4993,  -569.6699,    54.2848,  -571.0992,   135.8392,
            -93.6641,    11.7999,   887.9709,   -41.1330,   467.5546]],

        [[   20.5619,  -180.5011,  -372.7271,  1948.1963,   815.7861,
          -1230.3558,    13.7859,   128.2355,   -71.1078,   158.7931],
         [   17.4993,  -569.6699,    54.2848,  -571.0992,   135.8392,
            -93.6641,    11.7999,   887.9709,   -41.1330,   467.5546]],

        [[   20.5619,  -180.5011,  -372.7271,  1948.1963,   815.7861,
          -1230.3558,    13.7859,   128.2355,   -71.1078,   158.7931],
         [   17.4993,  -569.6699,    54.2848,  -571.0992,   135.8392,
            -93.6641,    11.7999,   887.9709,   -41.1330,   467.5546]],

        [[   20.5619,  -180.5011,  -372.7271,  1948.1981,   815.7857,
          -1230.3500,    13.7859,   128.2355,   -71.1078,   158.7931],
         [   17.4993,  -569.6702,    54.2848,  -571.1004,   135.8392,
            -93.6641,    11.7999,   887.9719,   -41.1330,   467.5547]]])
Abs diff max: 1.922959e+00, mean: 1.902504e-01
Rel diff max: 9.880201e-04, mean: 1.823883e-04
Inconsistent: 90 / 200

```

### Versions

PyTorch version: 2.10.0+cu126

OS: Ubuntu 24.04.1 LTS

GPU: NVIDIA GeForce RTX 4060 Laptop GPU

CUDA version: 12.6

Python version: 3.12.3

cc @chauhang @penguinwu

**Comment by malfet** [comment]
Closing as won't-fix.

The repro compares an **eager float64** run against a **compiled float32** run (`model.double()` vs `model_fp32.float()`) and flags any deviation above a hand-picked `abs > 0.01 | rel > 0.001`. That threshold is tighter than float32 can deliver through an LSTM + multi-head-attention chain, and the report gives no justification for why the float32 result is *wrong* rather than simply lower-precision. As the bug template asks: "If the difference is not expected as described in these documents, please provide appropriate justification on why one result is wrong and the other is correct." See [Numerical accuracy](https://docs.pytorch.org/docs/stable/notes/numerical_accuracy.html): "PyTorch is not guaranteed to produce bitwise identical results for floating point computations that are mathematically identical."

The reported magnitude is an amplification artifact rather than evidence of a compiler bug. The model applies `torch.reciprocal` three times to attention outputs that pass close to zero. Near zero, `1/x` turns an ordinary float32 rounding difference in `x` into an arbitrarily large difference in the output — which is why a max absolute difference of 1.92 appears alongside eager values already in the ±2000 range, i.e. a small *relative* error.

To show a genuine `torch.compile` correctness issue here, run both paths in the same dtype (float32 eager vs float32 compiled, same weights and inputs) and compare at default `rtol`/`atol`. #188425 was closed for the same reason.
